### PR TITLE
Switch from scalaz to cats (closes #61)

### LINKED ISSUE #61: Switch from scalaz to cats

### PR DESCRIPTION
**Pull request by yeonhoo**

### CODE REVIEW

**Comment by snowplowcla** [comment]
Confirmed! @yeonhoo has signed the Individual Contributor License Agreement. Thanks so much

File: .gitignore [hunk 1]
```diff
@@ -11,3 +11,5 @@ project/plugins/project/
 
 # Scala-IDE specific
 .scala_dependencies
+
+.idea
```
Comment (BenFradet) [inline .gitignore]: we usually don't have editor-specific stuff in the gitignore, the above actually prompted me to log #75

File: src/main/scala/com.snowplowanalytics.maxmind.iplookups/IpLookups.scala [hunk 2]
```diff
@@ -51,6 +52,11 @@ object IpLookups {
       memCache,
       lruCache
     )
+
+  implicit class RichValidated[E, A](validated: Validated[E, A]) {
+    def flatMap[EE >: E, B](f: A => Validated[EE, B]): Validated[EE, B] =
+      validated.andThen(f)
+  }
```
Comment (BenFradet) [inline src/main/scala/com.snowplowanalytics.maxmind.iplookups/IpLookups.scala]: I invite you to read https://typelevel.org/cats/datatypes/validated.html#of-flatmaps-and-eithers on why `flatMap` on `Validated` is not a good idea.

The problem is that it was defined on Scalaz' `Validation` and we've been abusing it a lot throughout the the different codebases.

Here we don't actually need the error-accumulating behaviour of `Validated` so I think we can safely move to `Either`.

File: .gitignore (hunk 1 above)
Comment (yeonhoo) [reply .gitignore]: Checked!

File: src/main/scala/com.snowplowanalytics.maxmind.iplookups/IpLookups.scala (hunk 2 above)
Comment (yeonhoo) [reply src/main/scala/com.snowplowanalytics.maxmind.iplookups/IpLookups.scala]: Ok, checked! I'm going to read now! Thank you Ben

**Comment by yeonhoo** [comment]
I've created an Error object where I defined all the errors but I realized now that all the errors are expected errors, right?

File: src/main/scala/com.snowplowanalytics.maxmind.iplookups/Errors.scala [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+      with IpLookupError
+
+  //com.maxmind.geoip2.exception.AddressNotFoundException is a final class
+  case class AddressNotFoundException(ex: com.maxmind.geoip2.exception.AddressNotFoundException)
+      extends IpLookupError {
+    val message: String   = ex.getMessage
+    override def toString = s"IpLookup ${ex.getClass.getSimpleName} " + message
+  }
+
+}
```
Comment (BenFradet) [inline src/main/scala/com.snowplowanalytics.maxmind.iplookups/Errors.scala]: The error ADT is a good idea but out of scope for this issue I think

File: src/main/scala/com.snowplowanalytics.maxmind.iplookups/IpLookups.scala [hunk 4]
```diff
@@ -190,7 +195,11 @@ class IpLookups(
...
-  /** Transforms a String into an Validation[Throwable, InetAddress] */
-  private def getIpAddress(ip: String): Validation[Throwable, InetAddress] =
-    Validation.fromTryCatch(InetAddress.getByName(ip))
+  /** Transforms a String into an Either[IpLookupError, InetAddress] */
+  private def getIpAddress(ip: String): Either[IpLookupError, InetAddress] =
+    try {
+      Right(InetAddress.getByName(ip))
+    } catch {
+      case ex: java.net.UnknownHostException => Left(UnknownHostException(ex.getMessage))
+    }
```
Comment (BenFradet) [inline src/main/scala/com.snowplowanalytics.maxmind.iplookups/IpLookups.scala]: you can just do [Either.catchNonFatal(...)](https://github.com/typelevel/cats/blob/master/core/src/main/scala/cats/syntax/either.scala#L306)

File: src/main/scala/com.snowplowanalytics.maxmind.iplookups/IpLookups.scala [hunk 5]
```diff
@@ -139,23 +136,31 @@ class IpLookups(
...
-          v   <- Validation.fromTryCatch(gs.city(ipA))
+          ipA <- ipAddress.right
+
+          v <- (try {
+            Right(gs.city(ipA))
+          } catch {
+            case ex: IOException                    => Left(ex)
+            case ex: GeoIp2AddressNotFoundException => Left(AddressNotFoundException(ex))
+            case ex: GeoIp2Exception                => Left(ex)
+          }).right
```
Comment (BenFradet) [inline src/main/scala/com.snowplowanalytics.maxmind.iplookups/IpLookups.scala]: same here

File: src/main/scala/com.snowplowanalytics.maxmind.iplookups/SpecializedReader.scala [hunk 6]
```diff
@@ -15,12 +15,20 @@ package com.snowplowanalytics.maxmind.iplookups
...
-  def getValue(ip: InetAddress): Validation[Throwable, String] =
-    Validation.fromTryCatch(f(db, ip))
+  def getValue(ip: InetAddress): Either[IpLookupError, String] = {
+    try {
+      Right(f(db, ip))
+    } catch {
+      case ex: IOException                    => Left(ex)
+      case ex: GeoIp2AddressNotFoundException => Left(AddressNotFoundException(ex))
+      case ex: GeoIp2Exception                => Left(ex)
+    }
```
Comment (BenFradet) [inline src/main/scala/com.snowplowanalytics.maxmind.iplookups/SpecializedReader.scala]: same

File: src/main/scala/com.snowplowanalytics.maxmind.iplookups/IpLookups.scala [hunk 7]
```diff
@@ -139,23 +136,31 @@ class IpLookups(
     val ipAddress = getIpAddress(ip)
 
     /**
-     * Creates a Validation boxing the result of using a lookup service on the ip
+     * Creates a Validated boxing the result of using a lookup service on the ip
```
Comment (BenFradet) [inline src/main/scala/com.snowplowanalytics.maxmind.iplookups/IpLookups.scala]: Either

File: README.md [hunk 8]
```diff
@@ -110,11 +110,11 @@ The `performLookups(ip)` method returns a:
...
-  ipLocation: Option[Validation[Throwable, IpLocation]],
-  isp: Option[Validation[Throwable, String]],
-  organization: Option[Validation[Throwable, String]],
-  domain: Option[Validation[Throwable, String]],
-  connectionType: Option[Validation[Throwable, String]]
+  ipLocation: Option[Validated[Throwable, IpLocation]],
+  isp: Option[Validated[Throwable, String]],
+  organization: Option[Validated[Throwable, String]],
+  domain: Option[Validated[Throwable, String]],
+  connectionType: Option[Validated[Throwable, String]]
```
Comment (BenFradet) [inline README.md]: Don't forget to change this too

File: src/main/scala/com.snowplowanalytics.maxmind.iplookups/IpLookups.scala [hunk 9]
```diff
@@ -18,9 +18,8 @@ import java.net.InetAddress
 import com.maxmind.db.CHMCache
 import com.maxmind.geoip2.DatabaseReader
 import com.twitter.util.SynchronizedLruMap
-import scalaz._
-
 import model._
+import cats.implicits._
```
Comment (BenFradet) [inline src/main/scala/com.snowplowanalytics.maxmind.iplookups/IpLookups.scala]: you can just do `cats.syntax.either._` and you can add it to the imports above, the second group is dedicated to the ones related to snowplow

File: src/main/scala/com.snowplowanalytics.maxmind.iplookups/SpecializedReader.scala [hunk 10]
```diff
@@ -16,11 +16,11 @@ import java.net.InetAddress
 
 import com.maxmind.geoip2.DatabaseReader
 import com.snowplowanalytics.maxmind.iplookups.ReaderFunctions.ReaderFunction
-import scalaz._
+import cats.implicits._
```
Comment (BenFradet) [inline src/main/scala/com.snowplowanalytics.maxmind.iplookups/SpecializedReader.scala]: same with the import

File: src/test/scala/com.snowplowanalytics.maxmind.iplookups/IpLookupsTest.scala [hunk 11]
```diff
@@ -12,23 +12,20 @@
...
 package com.snowplowanalytics.maxmind.iplookups
 
-import java.net.UnknownHostException
-
 import com.maxmind.geoip2.exception.AddressNotFoundException
+import java.net.UnknownHostException
 import org.specs2.mutable.Specification
-import scalaz._
-import Scalaz._
+import cats.implicits._
```
Comment (BenFradet) [inline src/test/scala/com.snowplowanalytics.maxmind.iplookups/IpLookupsTest.scala]: same

File: src/test/scala/com.snowplowanalytics.maxmind.iplookups/IpLookupsTest.scala [hunk 12]
```diff
@@ -42,67 +39,70 @@ object IpLookupsTest {
...
+          countryName = "China",
+          region = Some("22"),
+          city = Some("Changchun"),
+          latitude = 43.88F,
+          longitude = 125.3228F,
+          timezone = Some("Asia/Harbin"),
+          postalCode = None,
+          metroCode = None,
+          regionName = Some("Jilin Sheng")
+        )).some,
```
Comment (BenFradet) [inline src/test/scala/com.snowplowanalytics.maxmind.iplookups/IpLookupsTest.scala]: you can do `.right.some` and `.left.some` as was done before

File: src/main/scala/com.snowplowanalytics.maxmind.iplookups/IpLookups.scala (hunk 9 above)
Comment (yeonhoo) [reply src/main/scala/com.snowplowanalytics.maxmind.iplookups/IpLookups.scala]: Oh I see

File: src/test/scala/com.snowplowanalytics.maxmind.iplookups/IpLookupsTest.scala [hunk 13]
```diff
@@ -71,11 +69,11 @@ object IpLookupsTest {
...
-      ).success.some,
-      "Century Link".success.some,
-      "Lariat Software".success.some,
-      new AddressNotFoundException("The address 216.160.83.56 is not in the database.").failure.some,
-      new AddressNotFoundException("The address 216.160.83.56 is not in the database.").failure.some
+      ).asRight.some,
+      "Century Link".asRight.some,
+      "Lariat Software".asRight.some,
+      new AddressNotFoundException("The address 216.160.83.56 is not in the database.").asLeft.some,
+      new AddressNotFoundException("The address 216.160.83.56 is not in the database.").asLeft.some
```
Comment (BenFradet) [inline src/test/scala/com.snowplowanalytics.maxmind.iplookups/IpLookupsTest.scala]: why not `.right` and `.left`?

File: src/main/scala/com.snowplowanalytics.maxmind.iplookups/IpLookups.scala [hunk 14]
```diff
@@ -14,11 +14,10 @@ package com.snowplowanalytics.maxmind.iplookups
 
 import java.io.File
 import java.net.InetAddress
-
```
Comment (BenFradet) [inline src/main/scala/com.snowplowanalytics.maxmind.iplookups/IpLookups.scala]: yes sorry if my explanation wasn't complete earlier but the imports are grouped in this manner: jdk, other libraries, our libraries

File: src/test/scala/com.snowplowanalytics.maxmind.iplookups/IpLookupsTest.scala (hunk 13 above)
Comment (yeonhoo) [reply src/test/scala/com.snowplowanalytics.maxmind.iplookups/IpLookupsTest.scala]: there is no method `.right` and `.left`. or am I missing something?

File: src/main/scala/com.snowplowanalytics.maxmind.iplookups/IpLookups.scala (hunk 14 above)
Comment (yeonhoo) [reply src/main/scala/com.snowplowanalytics.maxmind.iplookups/IpLookups.scala]: oh right I saw now from other project's imports how imports are structured :+1:

File: src/test/scala/com.snowplowanalytics.maxmind.iplookups/IpLookupsTest.scala (hunk 13 above)
Comment (BenFradet) [reply src/test/scala/com.snowplowanalytics.maxmind.iplookups/IpLookupsTest.scala]: you're right my bad, I'm too used to scalaz

File: src/main/scala/com.snowplowanalytics.maxmind.iplookups/IpLookups.scala [hunk 15]
```diff
@@ -139,23 +135,23 @@ class IpLookups(
...
      * @return the result of the lookup
      */
-    def getLookup(service: Option[SpecializedReader]): Option[Validation[Throwable, String]] =
+    def getLookup(service: Option[SpecializedReader]): Option[Either[Throwable, String]] =
       service.map { s =>
         for {
-          ipA <- ipAddress
-          v   <- s.getValue(ipA)
+          ipA <- ipAddress.right
+          v   <- s.getValue(ipA).right
```
Comment (BenFradet) [inline src/main/scala/com.snowplowanalytics.maxmind.iplookups/IpLookups.scala]: why are the right projections needed?

**Pull request merged**